docs/arch/adr-003-outcome-centric-logic.md

```diff
@@ -0,0 +1,112 @@
+
+This approach:
+
+1. defines the required conditions on a per-outcome basis, so it's much easier to reason about how to get to a particular outcome
+1. flattens the heavily nested and complex structure, making it easier to make changes to the logic.
+
+If we look at the vast majority of Smart Answer flows which do _not_ suffer from this spaghetti code, such as the Register a Birth Smart Answer, the logic [is very easy to follow](https://github.com/alphagov/smart-answers/blob/7cc0ca0c85fe9de04d859e705dbe9b9b5d402306/lib/smart_answer_flows/register-a-birth.rb). Note that for each 'node', there is only one branch of logic to arrive at a particular outcome, and there is no nesting. In other words, **most Smart Answers already use an outcome-centric approach**.
+
+This proposal is to bring _all_ Smart Answers in line, by following these rules:
+
+```
+for every smart answer flow:
+  for every node in the flow:
+    - each outcome is only defined once
+    - there is only one level of indentation
+      (i.e. if/else and case statements are allowed, but no nesting)
+```
+
+This should bring all of our Smart Answers up to roughly the same readability standard, and we'll then have this ADR we can point our developers to, to prevent this from happening again.
+
+### Suggested steps for implementation
+
+Convert `pay-leave-for-parents.rb`, `check-uk-visa.rb` and `uk-benefits-abroad.rb` to be outcome-centric rather than input-centric. These are the worst offenders and would benefit most from the refactor.
+
+Then there are some small tweaks required to other smart answer flows - see 'Example conversion' below.
+
+Once all Smart Answers have been refactored to the same standard, there are [additional improvements we could consider](#additional-improvements). These are outside of the scope of this ADR, but are included for completion.
+
+### Example conversion
+
+Consider this block of code from [student-finance-calculator](https://github.com/alphagov/smart-answers/blob/7cc0ca0c85fe9de04d859e705dbe9b9b5d402306/lib/smart_answer_flows/student-finance-calculator.rb#L155-L179):
+
+```ruby
+next_node do |response|
+  case calculator.course_type
+    when "uk-full-time"
+      if response == "dental-medical-healthcare"
+        question :are_you_a_doctor_or_dentist?
+      else
+        outcome :outcome_uk_full_time_students
+      end
+    when "uk-part-time"
+      outcome :outcome_uk_all_students
+    else
+      outcome :outcome_eu_students
+    end
+end
+```
+
+To follow the proposal of this ADR, we'd change the code to this:
+
+```ruby
+next_node do |response|
+  if calculator.course_type == "uk-full-time" && response == "dental-medical-healthcare"
+    question :are_you_a_doctor_or_dentist?
+  elsif calculator.course_type == "uk-full-time"
+    outcome :outcome_uk_full_time_students
+  elsif calculator.course_type == "uk-part-time"
+    outcome :outcome_uk_all_students
+  else
+    outcome :outcome_eu_students
```

Comment:
With the new proposed rule, this code would look like:

```suggestion
  if calculator.should_ask_if_doctor_or_dentist?
    question :are_you_a_doctor_or_dentist?
  elsif calculator.uk_full_time_student?
    outcome :outcome_uk_full_time_students
  elsif calculator.uk_part_time_student?
    outcome :outcome_uk_all_students
  else
    outcome :outcome_eu_students
```

There is the question of what we should name these calculator methods - I note in your initial example you've [named the method the same as the outcome](https://github.com/alphagov/smart-answers/pull/5347#issuecomment-833439172), which read quite confusingly to me, though would probably help keep things consistent. I've proposed a more imperative naming style above, which feels a bit more natural and unit-testable. Would be interested in your thoughts.